src/compiler/mir/typetable.rs

```diff
@@ -0,0 +1,584 @@
+//! The system for keeping track of and defining types in the Bramble MIR.
+
+use std::fmt::Display;
+
+use crate::{
+    compiler::{
+        ast::{Node, Parameter, Path, PointerMut, StructDef, Type},
+        semantics::semanticnode::SemanticContext,
+    },
+    StringId,
+};
+
+pub struct TypeTable {
+    table: Vec<MirTypeDef>,
+}
+
+impl TypeTable {
+    pub fn new() -> TypeTable {
+        let table = vec![
+            MirTypeDef::Base(MirBaseType::Unit),
+            MirTypeDef::Base(MirBaseType::Null),
+            MirTypeDef::Base(MirBaseType::U8),
+            MirTypeDef::Base(MirBaseType::U16),
+            MirTypeDef::Base(MirBaseType::U32),
+            MirTypeDef::Base(MirBaseType::U64),
+            MirTypeDef::Base(MirBaseType::I8),
+            MirTypeDef::Base(MirBaseType::I16),
+            MirTypeDef::Base(MirBaseType::I32),
+            MirTypeDef::Base(MirBaseType::I64),
+            MirTypeDef::Base(MirBaseType::F64),
+            MirTypeDef::Base(MirBaseType::Bool),
+            MirTypeDef::Base(MirBaseType::StringLiteral),
+        ];
+
+        TypeTable { table }
+    }
+
+    pub fn get(&self, id: TypeId) -> &MirTypeDef {
+        &self.table[id.0 as usize]
+    }
+
+    fn get_mut(&mut self, id: TypeId) -> &mut MirTypeDef {
+        &mut self.table[id.0 as usize]
+    }
+
+    /// Adds the given [`Type`] to the type table. If this type references any type which is
+    /// not in the table then it will also add the referenced type to the table. Structures will
+    /// be added as [`MirStructDef::Declared`].
```

Comment:
Note that if the Type is already in the table then this will just return the ID of that occurrence.